addons/survey/controllers/main.py

```diff
@@ -728,39 +723,137 @@ def _get_user_input_domain(self, survey, line_filter_domain, **post):
             user_input_domain = expression.AND([[('scoring_success', '=', True)], user_input_domain])
-
         return user_input_domain
 
     def _extract_filters_data(self, survey, post):
+        """
+            Extracts the filters from the URL to returns the related user_input_lines and
+            the parameters used to render/remove the filters on the results page (search_filters).
+
+            :rtype: (recordset, List[dict])
+        """
+        filters_domains = []
         search_filters = []
-        line_filter_domain, line_choices = [], []
-        for data in post.get('filters', '').split('|'):
-            try:
-                row_id, answer_id = (int(item) for item in data.split(','))
-            except:
-                pass
-            else:
-                if row_id and answer_id:
-                    line_filter_domain = expression.AND([
-                        ['&', ('matrix_row_id', '=', row_id), ('suggested_answer_id', '=', answer_id)],
-                        line_filter_domain
-                    ])
-                    answers = request.env['survey.question.answer'].browse([row_id, answer_id])
-                elif answer_id:
-                    line_choices.append(answer_id)
-                    answers = request.env['survey.question.answer'].browse([answer_id])
-                if answer_id:
-                    question_id = answers[0].matrix_question_id or answers[0].question_id
-                    search_filters.append({
-                        'row_id': row_id,
-                        'answer_id': answer_id,
-                        'question': question_id.title,
-                        'answers': '%s%s' % (answers[0].value, ': %s' % answers[1].value if len(answers) > 1 else '')
-                    })
-        if line_choices:
-            line_filter_domain = expression.AND([[('suggested_answer_id', 'in', line_choices)], line_filter_domain])
 
-        user_input_domain = self._get_user_input_domain(survey, line_filter_domain, **post)
-        user_input_lines = request.env['survey.user_input'].sudo().search(user_input_domain).mapped('user_input_line_ids')
+        answer_cols_to_rows_mapping, answer_user_input_line_ids = self._get_filters_from_post(post)
+
+        # Matrix, Multiple choice, Simple choice filters
+        if answer_cols_to_rows_mapping:
+            ans_ids, row_ids = [], []
+            for filter_ans_id, filter_row_ids in answer_cols_to_rows_mapping.items():
+                ans_ids.append(filter_ans_id)
+                row_ids += filter_row_ids
+
+            answers = request.env['survey.question.answer'].browse(list(filter(bool, ans_ids+row_ids)))  # exclude '0'
+            # For performance, caches all useful columns of survey_question_answer in first query
+            answers_only = answers.filtered(lambda a: not a.matrix_question_id)   # not row_ids
+
+            for answer in answers_only:
+                for row_id in answer_cols_to_rows_mapping[answer.id]:
```

Comment:
Is there a test specifically checking the case of several row ids for the same answer_id?